src/compiler/checker.ts

```diff
@@ -6324,7 +6324,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
         else if (
             entityName.kind === SyntaxKind.QualifiedName || entityName.kind === SyntaxKind.PropertyAccessExpression ||
-            entityName.parent.kind === SyntaxKind.ImportEqualsDeclaration
+            entityName.parent.kind === SyntaxKind.ImportEqualsDeclaration ||
+            (entityName.parent.kind === SyntaxKind.QualifiedName && (entityName.parent as QualifiedName).left === entityName) ||
+            (entityName.parent.kind === SyntaxKind.PropertyAccessExpression && (entityName.parent as PropertyAccessExpression).expression === entityName) ||
+            (entityName.parent.kind === SyntaxKind.ElementAccessExpression && (entityName.parent as ElementAccessExpression).expression === entityName)

```

Comment:
These are so if you're the `A` in `A.B`, you get a `Namespace` meaning. Since we always get the leftmost identifier for the lookup, we only ever actually lookup the `A`. Even in an `A.B.C`, we only lookup the `A`. Elsewhere, we use `getQualifiedLeftMeaning` on the "input" meaning for the whole dotted name to map type/value meaning to a namespace/value meaning, but when we're using this function, we're usually in a state where we don't even have an "input" meaning yet.

Reply:
That this codepath is somewhat bespoke is... unfortunate. It basically has to distill "what meaning is done by a `resolveName` call where elsewhere in the checker for this node" into a single function, so declaration emit can redo the lookup with alias collection on.